File: trustly-android-sdk/src/main/java/net/trustly/android/sdk/views/clients/TrustlyWebViewChromeClient.java
package net.trustly.android.sdk.views.clients;

import android.os.Message;
import android.webkit.WebChromeClient;
import android.webkit.WebView;

import net.trustly.android.sdk.views.TrustlyView;

/**
 * Chrome client for Trustly View
 */
public class TrustlyWebViewChromeClient extends WebChromeClient {

    private final TrustlyView trustlyView;

    public TrustlyWebViewChromeClient(TrustlyView trustlyView) {
        this.trustlyView = trustlyView;
    }

    @Override
    public boolean onCreateWindow(WebView view, boolean isDialog, boolean isUserGesture, Message resultMsg) {
        return trustlyView.handleWebChromeClientOnCreateWindow(view, resultMsg);
    }

}

File: trustly-android-sdk/src/main/java/net/trustly/android/sdk/views/clients/TrustlyWebViewChromeClient.kt
package net.trustly.android.sdk.views.clients

import android.content.Context
import android.os.Message
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebView.WebViewTransport
import net.trustly.android.sdk.views.TrustlyView
import net.trustly.android.sdk.views.events.TrustlyEvents
import net.trustly.android.sdk.views.oauth.TrustlyOAuthView

/**
 * Chrome client for Trustly View
 */
class TrustlyWebViewChromeClient(
    private val context: Context,
    private val trustlyView: TrustlyView,
    private val trustlyEvents: TrustlyEvents,
) : WebChromeClient() {

    override fun onCreateWindow(
        view: WebView,
        isDialog: Boolean,
        isUserGesture: Boolean,
        resultMsg: Message
    ) = handleWebChromeClientOnCreateWindow(view, resultMsg)

    private fun handleWebChromeClientOnCreateWindow(view: WebView, resultMsg: Message): Boolean {
        val result = view.hitTestResult
        return if (result.type == 0) {
            //window.open
            val trustlyOAuthView = TrustlyOAuthView(context)
            trustlyView.addView(trustlyOAuthView)
            val transport = resultMsg.obj as WebViewTransport
            transport.webView = trustlyOAuthView.getWebView()
            resultMsg.sendToTarget()
            true
        } else {
            val url = result.extra
            url?.let {
                val params = mapOf("url" to url)
                trustlyEvents.handleOnExternalUrl(trustlyView, params)
            }
            false
        }
    }

}